utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala

```diff
@@ -64,30 +64,46 @@ object BroadcastUtils {
    * Returns a UDF that takes values of join keys and returns the value of the target attribute or null
    * if there is no mapping for the specified keys.
    *
-   * @param mappingTable        A mapping table broadcasted to executors.
-   * @param defaultValueExprMap The map of defaults with string expressions
+   * @param mappingTable            A mapping table broadcasted to executors.
+   * @param defaultValueExpressions The map of defaults with string expressions
    * @return A UDF that maps joins keys to a target attribute value.
    */
   def getMappingUdfForSingleOutput(mappingTable: Broadcast[LocalMappingTable],
-                                   defaultValueExprMap: Map[String, String])(implicit spark: SparkSession): UserDefinedFunction = {
+                                   defaultValueExpressions: Map[String, String])(implicit spark: SparkSession): UserDefinedFunction = {
     val numberOfArguments = mappingTable.value.keyTypes.size
 
-    val defaultValueMap = defaultValueExprMap.mapValues(getValueOfSparkExpression)
+    val defaultValue = defaultValueExpressions.mapValues(getValueOfSparkExpression)
+      .get(mappingTable.value.outputColumns.values.head).orNull
 
-    val lambda = getMappingLambda(mappingTable, defaultValueMap, numberOfArguments, isSingleOutput = true)
-
-    lambda
+    getLambda(numberOfArguments, mappingTable.value.getRowWithDefault(_, defaultValue), mappingTable.value.valueTypes.head)
   }
 
+  /**
+   * Returns a UDF that takes values of join keys and returns the target attribute values as a row or null
+   * if there is no mapping for the specified keys.
+   *
+   * @param mappingTable            A mapping table broadcasted to executors.
+   * @param defaultValueExpressions The map of defaults with string expressions
+   * @return A UDF that maps joins keys to a row of target attribute values.
+   */
   def getMappingUdfForMultipleOutputs(mappingTable: Broadcast[LocalMappingTable],
-                                   defaultValueExprMap: Map[String, String])(implicit spark: SparkSession): UserDefinedFunction = {
+                                      defaultValueExpressions: Map[String, String])(implicit spark: SparkSession): UserDefinedFunction = {
     val numberOfArguments = mappingTable.value.keyTypes.size
 
-    val defaultValueMap = defaultValueExprMap.mapValues(getValueOfSparkExpression)
+    val defaultValues = defaultValueExpressions.mapValues(getValueOfSparkExpression)
+    val defaultRow = if (defaultValues.isEmpty) {
+      null
+    } else {
+      Row.fromSeq(mappingTable.value.outputColumns.values.toSeq.map(field => defaultValues.getOrElse(field, null)))
+    }
 
-    val lambda = getMappingLambda(mappingTable, defaultValueMap, numberOfArguments, isSingleOutput = false)
+    val fnc: Seq[Any] => Any = (seq: Seq[Any] ) => mappingTable.value.getRowWithDefault(seq, defaultRow)
+    val structFields: Seq[StructField] = mappingTable.value.outputColumns.keys
```

Comment:
This is one I wanted to suggest for refactoring during the original multicolumne result back then around Easter (but didn't happen for reasons 😉 ).
Anyway, I think it should be a property of `LocalMappingTable`, the returned `DataType` of the lookup. IMHO it belongs there conceptually, and might make the overall code tighter - construct and define the `returnedType: DataType` (name is just a suggestion) during `LocalMappingTable` creation.

Reply:
But how to do with the ErrorMessage value type?

Reply:
Ah, of course. After all `LocalMappingTable` is not its (`getLambda`) parameter anymore anyway.